TypeCobol.Analysis/Cfg/ControlFlowGraphBuilder.cs

```diff
@@ -0,0 +1,2695 @@
+            {
+                if (IsMatchingPath(candidate, paths))
+                    results.Add(candidate);
+            }
+            return results;
+
+            string[] AsPath(SymbolReference symbolReference)
+            {
+                if (symbolReference.IsQualifiedReference)
+                {
+                    var qualifiedSymbolReference = (QualifiedSymbolReference) symbolReference;
+                    return qualifiedSymbolReference.AsList().Select(s => s.Name).ToArray();
+                }
+                return new[] { symbolReference.Name };
+            }
+
+            bool IsMatchingPath(Symbol symbol, string[] path)
+            {
+                Symbol currentSymbol = symbol;
+                for (int i = 0; i < path.Length; i++)
+                {
+                    switch (i)
+                    {
+                        case 0:
+                            if (!path[i].Equals(currentSymbol.Name, StringComparison.OrdinalIgnoreCase))
+                                return false;
+                            break;
+                        default:
+                        {
+                            Symbol parent = currentSymbol.LookupParentOfName(path[i]);
+                            if (parent == null)
+                                return false;
+                            currentSymbol = parent;
+                        }
+                            break;
+                    }
+                }
+                return true;
+            }
+        }
+
+        /// <summary>
+        /// Enter a section declaration
+        /// </summary>
+        /// <param name="section"></param>
+        protected virtual void EnterSection(Section section)
+        {
+            string name = section.Name;
+            CfgSectionSymbol sym = new CfgSectionSymbol(name);
+            this.CurrentProgramCfgBuilder.EnterSectionOrParagraphSymbol(sym);
+            //The new current section.
+            this.CurrentProgramCfgBuilder.CurrentSection = sym;
+            //No more Paragraph
+            this.CurrentProgramCfgBuilder.CurrentParagraph = null;
+            //Reset any current sentence
+            this.CurrentProgramCfgBuilder.CurrentSentence = null;
+        }
+
+        /// <summary>
+        /// /Leave a section declaration.
+        /// </summary>
+        /// <param name="section"></param>
+        protected virtual void LeaveSection(Section section)
+        {
+            this.CurrentProgramCfgBuilder.CurrentSection = null;
+            //Current sentence is also null now
+            this.CurrentProgramCfgBuilder.CurrentSentence = null;
+        }
+
+        /// <summary>
+        /// Enter a paragraph
+        /// </summary>
+        /// <param name="p">The paragraph to be entered</param>
+        protected virtual void EnterParagraph(Paragraph p)
+        {
+            string name = p.Name;
+            CfgParagraphSymbol sym = new CfgParagraphSymbol(name);
```

Comment:
I think there are two ways to see this:
- either we consider the CFG is dependent from SD and then the control flow graph building is some sort of extension of the symbol building process
  - `ControlFlowGraphBuilder` has to know about the `ProgramSymbolTableBuilder` so it can do its own stuff while symbols are built
- or the two processes are independent of each other which means `ControlFlowGraphBuilder` only knows about Node building and never uses symbols
  - CFG builder uses its own representation of section and paragraphs which are not related to symbols

I personnally think the second way is cleaner because it would be simpler to change the SD or the CFG when they are independent (no need to worry about consequences in one feature when modifying the other)

Reply:
The second option seems fine for me.
I think ControlFlowGraphBuilder ~should~ must not know about ProgramSymbolTableBuilder.
Builders are internal implementation and are not "public".

Reply:
All things considered, I prefer the second option:

> or the two processes are independent of each other which means ControlFlowGraphBuilder only knows about Node building and never uses symbols
CFG builder uses its own representation of section and paragraphs which are not related to symbols

CfgParagrahSymbol and CfgSectionSymbol should not inherent from Symbol classes but rather "attach" information to Symbol.
I see 2 ways to do this:
- Manage a Dictionary in CFG to map a Symbol to additional info
- Add a Dictionary<string, object> into Symbol class. string = free identifier, object = information to attach. This dictionary can be use by any class to "attach" information to a Symbol.

Reply:
CFG now uses its own classes for sentences, sections and paragraphs and they are not related to SemanticDomain.